numalogic/udfs/trainer.py

```diff
@@ -0,0 +1,330 @@
+            threshold_clf=thresh_clf,
+            trainer_cfg=_conf.numalogic_conf.trainer,
+        )
+
+        # Save artifacts
+        # TODO perform multi-save here
+        self.save_artifact(
+            artifacts["preproc_clf"],
+            skeys=payload.composite_keys,
+            dkeys=[_conf.name for _conf in _conf.numalogic_conf.preprocess],
+            uuid=payload.uuid,
+        )
+        self.save_artifact(
+            artifacts["model"],
+            skeys=payload.composite_keys,
+            dkeys=[_conf.numalogic_conf.model.name],
+            uuid=payload.uuid,
+            train_size=x_train.shape[0],
+        )
+        self.save_artifact(
+            artifacts["threshold_clf"],
+            skeys=payload.composite_keys,
+            dkeys=[_conf.numalogic_conf.threshold.name],
+            uuid=payload.uuid,
+        )
+
+        _LOGGER.debug(
+            "%s - Time taken in trainer: %.4f sec", payload.uuid, time.perf_counter() - _start_time
+        )
+        return Messages(Message.to_drop())
+
+    @staticmethod
+    def _construct_preproc_clf(_conf: StreamConf) -> Optional[artifact_t]:
+        preproc_factory = PreprocessFactory()
+        preproc_clfs = []
+        for _cfg in _conf.numalogic_conf.preprocess:
+            _clf = preproc_factory.get_instance(_cfg)
+            preproc_clfs.append(_clf)
+        if not preproc_clfs:
+            return None
+        if len(preproc_clfs) == 1:
+            return preproc_clfs[0]
+        return make_pipeline(*preproc_clfs)
+
+    def save_artifact(
+        self, artifact: artifact_t, skeys: list[str], dkeys: list[str], uuid: str, **metadata
+    ) -> None:
+        """
+        Save artifact to the registry.
+
+        Args:
+            artifact: Artifact to save
+            skeys: List of keys
+            dkeys: List of dkeys
+            uuid: UUID
+            **metadata: Additional metadata
+        """
+        if not artifact:
+            return
+        # TODO check for statelessness from config
+        if isinstance(artifact, StatelessTransformer):
+            _LOGGER.info("%s - Skipping save for stateless artifact with dkeys: %s", uuid, dkeys)
+            return
+        try:
+            version = self.model_registry.save(
+                skeys=skeys,
+                dkeys=dkeys,
+                artifact=artifact,
+                uuid=uuid,
+                **metadata,
+            )
+        except RedisRegistryError:
+            _LOGGER.exception("%s - Error while saving Model with skeys: %s", uuid, skeys)
+        else:
+            _LOGGER.info(
+                "%s - Artifact saved with dkeys: %s with version: %s", uuid, dkeys, version
+            )
+
+    def _is_data_sufficient(self, payload: TrainerPayload, df: pd.DataFrame) -> bool:
+        _conf = self.get_conf(payload.config_id)
+        return len(df) > _conf.numalogic_conf.trainer.min_train_size
+
+    def _is_new_request(self, payload: TrainerPayload) -> bool:
```

Comment:
this needs be improved. we need to look in to redis after deduping to check if it has a latest model. (later)